<a href="https://colab.research.google.com/github/Arooj123-bit/Python-Pandas/blob/main/Pandas_Day1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### In order to use any library in python, you need to import that library in you notebook

## Import Pandas

In [ ]:
import pandas as pd # alias pandas as pd
# import library name
# importing library will give me an acces to all feature in this library
# this library is only avaliable in this notebook.

In [ ]:
data = {
    "calories": [420, 380, 390],
    "duration": [50, 40,45]
}
# this is a dictionary
# using DataFrame,  we will convert dict to DataFrame
pd.DataFrame(data)
# this code: go to pandas library, use dataframe function and turn data dict into dataframe

calories  duration
0       420        50
1       380        40
2       390        45

###### Assign dataframe to df

In [ ]:
df = pd.DataFrame(data)
# this code will create a df variable

## Reading CSV or Excel File

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving student.xlsx to student.xlsx


In [ ]:
df= pd.read_excel("student.xlsx")

In [ ]:
df

calories  duration
0       420        50
1       380        40
2       390        45

In [ ]:
# first 5 rows
df.head()

id        name  class  mark  gender
0   1    John Deo   Four    75  female
1   2    Max Ruin  Three    85    male
2   3      Arnold  Three    55    male
3   4  Krish Star   Four    60  female
4   5   John Mike   Four    60  female

In [ ]:
df.head(3)

id      name  class  mark  gender
0   1  John Deo   Four    75  female
1   2  Max Ruin  Three    85    male
2   3    Arnold  Three    55    male

In [ ]:
df.head(10)

id         name  class  mark  gender
0   1     John Deo   Four    75  female
1   2     Max Ruin  Three    85    male
2   3       Arnold  Three    55    male
3   4   Krish Star   Four    60  female
4   5    John Mike   Four    60  female
5   6    Alex John   Four    55    male
6   7  My John Rob  Fifth    78    male
7   8       Asruid   Five    85    male
8   9      Tes Qry    Six    78     NaN
9  10     Big John   Four    55  female

In [ ]:
df.count()

id        35
name      34
class     34
mark      35
gender    33
dtype: int64

In [ ]:
df.sample()

id       name class  mark  gender
32  33  Kenn Rein   Six    96  female

In [ ]:
df.sample(5)

id         name  class  mark  gender
18  19        Tinny   Nine    18    male
6    7  My John Rob  Fifth    78    male
10  11       Ronald    Six    89  female
22  23        Herod  Eight    79    male
5    6    Alex John   Four    55    male

In [ ]:
df.shape

(35, 5)

In [ ]:
# dataset information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      35 non-null     int64 
 1   name    34 non-null     object
 2   class   34 non-null     object
 3   mark    35 non-null     int64 
 4   gender  33 non-null     object
dtypes: int64(2), object(3)
memory usage: 1.5+ KB


In [ ]:
# name of the columns
df.columns

Index(['id', 'name', 'class', 'mark', 'gender'], dtype='object')

In [ ]:
# statistical summary
df.describe()

id       mark
count  35.000000  35.000000
mean   18.000000  74.657143
std    10.246951  16.401117
min     1.000000  18.000000
25%     9.500000  62.500000
50%    18.000000  79.000000
75%    26.500000  88.000000
max    35.000000  96.000000

<p>std = standart deviation
<p>25% = 1st quarter
<p>50% = median values
<p>75% = 3rd quarter
<p>25% represent: 25% of the students have mark equal to or less than 62.5
<p>mean usually is higher than std. if std is close or higher than mean, there are outliers.

In [ ]:
#number of rows
len(df)

35

In [ ]:
# rows * column
# 35 * 5 = 175
df.size


175

# How to subset and Select data

In [ ]:
# return only name column
df["name"]

0        John Deo
1        Max Ruin
2          Arnold
3      Krish Star
4       John Mike
5       Alex John
6     My John Rob
7          Asruid
8         Tes Qry
9        Big John
10         Ronald
11          Recky
12            Kty
13           Bigy
14       Tade Row
15          Gimmy
16          Tumyu
17          Honny
18          Tinny
19         Jackly
20     Babby John
21         Reggid
22          Herod
23      Tiddy Now
24       Giff Tow
25         Crelea
26            NaN
27      Rojj Base
28    Tess Played
29      Reppy Red
30    Marry Toeey
31      Binn Rott
32      Kenn Rein
33       Gain Toe
34     Rows Noump
Name: name, dtype: object

In [ ]:
df["gender"]

0     female
1       male
2       male
3     female
4     female
5       male
6       male
7       male
8        NaN
9     female
10    female
11    female
12    female
13    female
14      male
15      male
16      male
17      male
18      male
19    female
20    female
21    female
22      male
23      male
24      male
25      male
26       NaN
27    female
28      male
29    female
30      male
31    female
32    female
33      male
34    female
Name: gender, dtype: object

In [ ]:
df.id

0      1
1      2
2      3
3      4
4      5
5      6
6      7
7      8
8      9
9     10
10    11
11    12
12    13
13    14
14    15
15    16
16    17
17    18
18    19
19    20
20    21
21    22
22    23
23    24
24    25
25    26
26    27
27    28
28    29
29    30
30    31
31    32
32    33
33    34
34    35
Name: id, dtype: int64

In [ ]:
df.mark

0     75
1     85
2     55
3     60
4     60
5     55
6     78
7     85
8     78
9     55
10    89
11    94
12    88
13    88
14    88
15    88
16    54
17    75
18    18
19    65
20    69
21    55
22    79
23    78
24    88
25    79
26    81
27    86
28    55
29    79
30    88
31    90
32    96
33    69
34    88
Name: mark, dtype: int64

In [ ]:
df.class
# this will return error, because class is reversed keyword

SyntaxError: invalid syntax (ipython-input-2999541473.py, line 1)

#### Select of multiple columns

In [ ]:
df[["name", "class"]]
# names and classes

name  class
0      John Deo   Four
1      Max Ruin  Three
2        Arnold  Three
3    Krish Star   Four
4     John Mike   Four
5     Alex John   Four
6   My John Rob  Fifth
7        Asruid   Five
8       Tes Qry    Six
9      Big John   Four
10       Ronald    Six
11        Recky    Six
12          Kty  Seven
13         Bigy  Seven
14     Tade Row    NaN
15        Gimmy   Four
16        Tumyu    Six
17        Honny   Five
18        Tinny   Nine
19       Jackly   Nine
20   Babby John   Four
21       Reggid  Seven
22        Herod  Eight
23    Tiddy Now  Seven
24     Giff Tow  Seven
25       Crelea  Seven
26          NaN  Three
27    Rojj Base  Seven
28  Tess Played  Seven
29    Reppy Red    Six
30  Marry Toeey   Four
31    Binn Rott  Seven
32    Kenn Rein    Six
33     Gain Toe  Seven
34   Rows Noump    Six

In [ ]:
df[["name", "class" , "gender"]].head()
# head method can be applied to selected dataframe

name  class  gender
0    John Deo   Four  female
1    Max Ruin  Three    male
2      Arnold  Three    male
3  Krish Star   Four  female
4   John Mike   Four  female

In [ ]:
df[["id", "mark"]].sample(3)

id  mark
23  24    78
12  13    88
22  23    79

### Lets Calculate average marks.

In [ ]:
# select mark column and applied mean
df["mark"].mean()

np.float64(74.65714285714286)

## Loc and iloc

<p>df.loc[rows, columns name(s)]
<p>df.iloc[rows, column index(es)]

In [ ]:
df.loc[0:34 ,"name"]
# return all rows but only name column

0        John Deo
1        Max Ruin
2          Arnold
3      Krish Star
4       John Mike
5       Alex John
6     My John Rob
7          Asruid
8         Tes Qry
9        Big John
10         Ronald
11          Recky
12            Kty
13           Bigy
14       Tade Row
15          Gimmy
16          Tumyu
17          Honny
18          Tinny
19         Jackly
20     Babby John
21         Reggid
22          Herod
23      Tiddy Now
24       Giff Tow
25         Crelea
26            NaN
27      Rojj Base
28    Tess Played
29      Reppy Red
30    Marry Toeey
31      Binn Rott
32      Kenn Rein
33       Gain Toe
34     Rows Noump
Name: name, dtype: object

In [ ]:
df.loc[0:4 ,"name"]
# return first 5 rows but only name column

0      John Deo
1      Max Ruin
2        Arnold
3    Krish Star
4     John Mike
Name: name, dtype: object

In [ ]:
df.loc[:, "name"]
# return all rows
# dont write start and end of the range :

0        John Deo
1        Max Ruin
2          Arnold
3      Krish Star
4       John Mike
5       Alex John
6     My John Rob
7          Asruid
8         Tes Qry
9        Big John
10         Ronald
11          Recky
12            Kty
13           Bigy
14       Tade Row
15          Gimmy
16          Tumyu
17          Honny
18          Tinny
19         Jackly
20     Babby John
21         Reggid
22          Herod
23      Tiddy Now
24       Giff Tow
25         Crelea
26            NaN
27      Rojj Base
28    Tess Played
29      Reppy Red
30    Marry Toeey
31      Binn Rott
32      Kenn Rein
33       Gain Toe
34     Rows Noump
Name: name, dtype: object

In [ ]:
df.loc[:10, "name"]
# return up to index 10
# dont write start and end of the range :

0        John Deo
1        Max Ruin
2          Arnold
3      Krish Star
4       John Mike
5       Alex John
6     My John Rob
7          Asruid
8         Tes Qry
9        Big John
10         Ronald
Name: name, dtype: object

In [ ]:
df.loc[10:, "name"]
# return from index 10
# dont write start and end of the range :

10         Ronald
11          Recky
12            Kty
13           Bigy
14       Tade Row
15          Gimmy
16          Tumyu
17          Honny
18          Tinny
19         Jackly
20     Babby John
21         Reggid
22          Herod
23      Tiddy Now
24       Giff Tow
25         Crelea
26            NaN
27      Rojj Base
28    Tess Played
29      Reppy Red
30    Marry Toeey
31      Binn Rott
32      Kenn Rein
33       Gain Toe
34     Rows Noump
Name: name, dtype: object

In [ ]:
df.loc[10:, ["name", "id"]]
# return from index 10
# dont write start and end of the range :

name  id
10       Ronald  11
11        Recky  12
12          Kty  13
13         Bigy  14
14     Tade Row  15
15        Gimmy  16
16        Tumyu  17
17        Honny  18
18        Tinny  19
19       Jackly  20
20   Babby John  21
21       Reggid  22
22        Herod  23
23    Tiddy Now  24
24     Giff Tow  25
25       Crelea  26
26          NaN  27
27    Rojj Base  28
28  Tess Played  29
29    Reppy Red  30
30  Marry Toeey  31
31    Binn Rott  32
32    Kenn Rein  33
33     Gain Toe  34
34   Rows Noump  35

In [ ]:
df.iloc[:, 1]
#1 is index for name column

0        John Deo
1        Max Ruin
2          Arnold
3      Krish Star
4       John Mike
5       Alex John
6     My John Rob
7          Asruid
8         Tes Qry
9        Big John
10         Ronald
11          Recky
12            Kty
13           Bigy
14       Tade Row
15          Gimmy
16          Tumyu
17          Honny
18          Tinny
19         Jackly
20     Babby John
21         Reggid
22          Herod
23      Tiddy Now
24       Giff Tow
25         Crelea
26            NaN
27      Rojj Base
28    Tess Played
29      Reppy Red
30    Marry Toeey
31      Binn Rott
32      Kenn Rein
33       Gain Toe
34     Rows Noump
Name: name, dtype: object

In [ ]:
df.iloc[:, [1,2]]
#1 is index for name column

name  class
0      John Deo   Four
1      Max Ruin  Three
2        Arnold  Three
3    Krish Star   Four
4     John Mike   Four
5     Alex John   Four
6   My John Rob  Fifth
7        Asruid   Five
8       Tes Qry    Six
9      Big John   Four
10       Ronald    Six
11        Recky    Six
12          Kty  Seven
13         Bigy  Seven
14     Tade Row    NaN
15        Gimmy   Four
16        Tumyu    Six
17        Honny   Five
18        Tinny   Nine
19       Jackly   Nine
20   Babby John   Four
21       Reggid  Seven
22        Herod  Eight
23    Tiddy Now  Seven
24     Giff Tow  Seven
25       Crelea  Seven
26          NaN  Three
27    Rojj Base  Seven
28  Tess Played  Seven
29    Reppy Red    Six
30  Marry Toeey   Four
31    Binn Rott  Seven
32    Kenn Rein    Six
33     Gain Toe  Seven
34   Rows Noump    Six

In [ ]:
# df.loc[0:4, ["name","class"]] first 5 rows
# in loc, stopping range is included in the output

df.iloc[0:5, [1,2]]
# in iloc, stopping range is excluded in the output

name  class
0    John Deo   Four
1    Max Ruin  Three
2      Arnold  Three
3  Krish Star   Four
4   John Mike   Four

In [ ]:
df.iloc[0:5, 0:2]
# 5 rows and 2 columns

id        name
0   1    John Deo
1   2    Max Ruin
2   3      Arnold
3   4  Krish Star
4   5   John Mike

## how to filter data

##### using slicing method df[condition]

In [ ]:
df[df["mark"]>90]


id       name class  mark  gender
11  12      Recky   Six    94  female
32  33  Kenn Rein   Six    96  female

In [ ]:
# return true or false
df["mark"]>90

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11     True
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
32     True
33    False
34    False
Name: mark, dtype: bool

In [ ]:
df[df["class"]=="Six"]

id        name class  mark  gender
8    9     Tes Qry   Six    78     NaN
10  11      Ronald   Six    89  female
11  12       Recky   Six    94  female
16  17       Tumyu   Six    54    male
29  30   Reppy Red   Six    79  female
32  33   Kenn Rein   Six    96  female
34  35  Rows Noump   Six    88  female

#### Filter by multiple condition

In [ ]:
df[(df["class"]=="Six") & (df["gender"]=="male")]

id   name class  mark gender
16  17  Tumyu   Six    54   male

In [ ]:
df[(df["gender"]=="female") & (df["mark"]>=85)]

id        name  class  mark  gender
10  11      Ronald    Six    89  female
11  12       Recky    Six    94  female
12  13         Kty  Seven    88  female
13  14        Bigy  Seven    88  female
27  28   Rojj Base  Seven    86  female
31  32   Binn Rott  Seven    90  female
32  33   Kenn Rein    Six    96  female
34  35  Rows Noump    Six    88  female

In [ ]:
# | this is called pipe
df[(df["gender"]=="female") | (df["mark"]>=90)]

id        name  class  mark  gender
0    1    John Deo   Four    75  female
3    4  Krish Star   Four    60  female
4    5   John Mike   Four    60  female
9   10    Big John   Four    55  female
10  11      Ronald    Six    89  female
11  12       Recky    Six    94  female
12  13         Kty  Seven    88  female
13  14        Bigy  Seven    88  female
19  20      Jackly   Nine    65  female
20  21  Babby John   Four    69  female
21  22      Reggid  Seven    55  female
27  28   Rojj Base  Seven    86  female
29  30   Reppy Red    Six    79  female
31  32   Binn Rott  Seven    90  female
32  33   Kenn Rein    Six    96  female
34  35  Rows Noump    Six    88  female

### df.loc[conditiona,column labels]

In [ ]:
df.loc[df["id"]<5, : ]


id        name  class  mark  gender
0   1    John Deo   Four    75  female
1   2    Max Ruin  Three    85    male
2   3      Arnold  Three    55    male
3   4  Krish Star   Four    60  female

In [ ]:
df.loc[df["gender"]=="male", ["name","class"]]

name  class
1      Max Ruin  Three
2        Arnold  Three
5     Alex John   Four
6   My John Rob  Fifth
7        Asruid   Five
14     Tade Row    NaN
15        Gimmy   Four
16        Tumyu    Six
17        Honny   Five
18        Tinny   Nine
22        Herod  Eight
23    Tiddy Now  Seven
24     Giff Tow  Seven
25       Crelea  Seven
28  Tess Played  Seven
30  Marry Toeey   Four
33     Gain Toe  Seven

In [ ]:
df.iloc[df["gender"]=="male", 1]
# iloc cant be used to filter

NotImplementedError: iLocation based boolean indexing on an integer type is not available

In [ ]:
df['mark'].mean()

np.float64(74.65714285714286)

In [ ]:
df.groupby('gender')['mark'].mean()


gender
female    77.312500
male      71.588235
Name: mark, dtype: float64

In [ ]:
df.loc[df["class"] == "Five", "mark"].mean()

np.float64(80.0)

In [ ]:
df["mark"].max()

96

In [ ]:
df["mark"].min()

18

In [ ]:
df.loc[(df["class"]=="Six") & (df["gender"]=="male"), "mark"].mean()

np.float64(54.0)

In [ ]:
# Sorting by names

df.loc[:,:].sort_values(by="name")

id         name  class  mark  gender
5    6    Alex John   Four    55    male
2    3       Arnold  Three    55    male
7    8       Asruid   Five    85    male
20  21   Babby John   Four    69  female
9   10     Big John   Four    55  female
13  14         Bigy  Seven    88  female
31  32    Binn Rott  Seven    90  female
25  26       Crelea  Seven    79    male
33  34     Gain Toe  Seven    69    male
24  25     Giff Tow  Seven    88    male
15  16        Gimmy   Four    88    male
22  23        Herod  Eight    79    male
17  18        Honny   Five    75    male
19  20       Jackly   Nine    65  female
0    1     John Deo   Four    75  female
4    5    John Mike   Four    60  female
32  33    Kenn Rein    Six    96  female
3    4   Krish Star   Four    60  female
12  13          Kty  Seven    88  female
30  31  Marry Toeey   Four    88    male
1    2     Max Ruin  Three    85    male
6    7  My John Rob  Fifth    78    male
11  12        Recky    Six    94  female
21  22       Reggid  Seven    55  female
29  30    Reppy Red    Six    79  female
27  28    Rojj Base  Seven    86  female
10  11       Ronald    Six    89  female
34  35   Rows Noump    Six    88  female
14  15     Tade Row    NaN    88    male
8    9      Tes Qry    Six    78     NaN
28  29  Tess Played  Seven    55    male
23  24    Tiddy Now  Seven    78    male
18  19        Tinny   Nine    18    male
16  17        Tumyu    Six    54    male
26  27          NaN  Three    81     NaN

In [ ]:
# Sorting by  descending order

df.loc[:,:].sort_values(by="name", ascending = False)

id         name  class  mark  gender
16  17        Tumyu    Six    54    male
18  19        Tinny   Nine    18    male
23  24    Tiddy Now  Seven    78    male
28  29  Tess Played  Seven    55    male
8    9      Tes Qry    Six    78     NaN
14  15     Tade Row    NaN    88    male
34  35   Rows Noump    Six    88  female
10  11       Ronald    Six    89  female
27  28    Rojj Base  Seven    86  female
29  30    Reppy Red    Six    79  female
21  22       Reggid  Seven    55  female
11  12        Recky    Six    94  female
6    7  My John Rob  Fifth    78    male
1    2     Max Ruin  Three    85    male
30  31  Marry Toeey   Four    88    male
12  13          Kty  Seven    88  female
3    4   Krish Star   Four    60  female
32  33    Kenn Rein    Six    96  female
4    5    John Mike   Four    60  female
0    1     John Deo   Four    75  female
19  20       Jackly   Nine    65  female
17  18        Honny   Five    75    male
22  23        Herod  Eight    79    male
15  16        Gimmy   Four    88    male
24  25     Giff Tow  Seven    88    male
33  34     Gain Toe  Seven    69    male
25  26       Crelea  Seven    79    male
31  32    Binn Rott  Seven    90  female
13  14         Bigy  Seven    88  female
9   10     Big John   Four    55  female
20  21   Babby John   Four    69  female
7    8       Asruid   Five    85    male
2    3       Arnold  Three    55    male
5    6    Alex John   Four    55    male
26  27          NaN  Three    81     NaN

In [ ]:
df[["name", "class", "mark"]].sort_values(by="class")

name  class  mark
22        Herod  Eight    79
6   My John Rob  Fifth    78
17        Honny   Five    75
7        Asruid   Five    85
30  Marry Toeey   Four    88
20   Babby John   Four    69
15        Gimmy   Four    88
9      Big John   Four    55
0      John Deo   Four    75
5     Alex John   Four    55
3    Krish Star   Four    60
4     John Mike   Four    60
19       Jackly   Nine    65
18        Tinny   Nine    18
12          Kty  Seven    88
33     Gain Toe  Seven    69
25       Crelea  Seven    79
31    Binn Rott  Seven    90
21       Reggid  Seven    55
28  Tess Played  Seven    55
23    Tiddy Now  Seven    78
24     Giff Tow  Seven    88
13         Bigy  Seven    88
27    Rojj Base  Seven    86
29    Reppy Red    Six    79
32    Kenn Rein    Six    96
34   Rows Noump    Six    88
8       Tes Qry    Six    78
16        Tumyu    Six    54
11        Recky    Six    94
10       Ronald    Six    89
1      Max Ruin  Three    85
26          NaN  Three    81
2        Arnold  Three    55
14     Tade Row    NaN    88

In [ ]:
df.loc[df["mark"]>50, :].sort_values(by="mark")

id         name  class  mark  gender
16  17        Tumyu    Six    54    male
2    3       Arnold  Three    55    male
9   10     Big John   Four    55  female
5    6    Alex John   Four    55    male
21  22       Reggid  Seven    55  female
28  29  Tess Played  Seven    55    male
3    4   Krish Star   Four    60  female
4    5    John Mike   Four    60  female
19  20       Jackly   Nine    65  female
20  21   Babby John   Four    69  female
33  34     Gain Toe  Seven    69    male
17  18        Honny   Five    75    male
0    1     John Deo   Four    75  female
8    9      Tes Qry    Six    78     NaN
6    7  My John Rob  Fifth    78    male
23  24    Tiddy Now  Seven    78    male
29  30    Reppy Red    Six    79  female
25  26       Crelea  Seven    79    male
22  23        Herod  Eight    79    male
26  27          NaN  Three    81     NaN
1    2     Max Ruin  Three    85    male
7    8       Asruid   Five    85    male
27  28    Rojj Base  Seven    86  female
12  13          Kty  Seven    88  female
24  25     Giff Tow  Seven    88    male
30  31  Marry Toeey   Four    88    male
15  16        Gimmy   Four    88    male
14  15     Tade Row    NaN    88    male
34  35   Rows Noump    Six    88  female
13  14         Bigy  Seven    88  female
10  11       Ronald    Six    89  female
31  32    Binn Rott  Seven    90  female
11  12        Recky    Six    94  female
32  33    Kenn Rein    Six    96  female

In [ ]:
# return student who are female or at class seven
df[(df["gender"]=="female") | (df["class"]=="Seven")].sort_values(by="mark")

id         name  class  mark  gender
9   10     Big John   Four    55  female
21  22       Reggid  Seven    55  female
28  29  Tess Played  Seven    55    male
3    4   Krish Star   Four    60  female
4    5    John Mike   Four    60  female
19  20       Jackly   Nine    65  female
20  21   Babby John   Four    69  female
33  34     Gain Toe  Seven    69    male
0    1     John Deo   Four    75  female
23  24    Tiddy Now  Seven    78    male
29  30    Reppy Red    Six    79  female
25  26       Crelea  Seven    79    male
27  28    Rojj Base  Seven    86  female
24  25     Giff Tow  Seven    88    male
12  13          Kty  Seven    88  female
13  14         Bigy  Seven    88  female
34  35   Rows Noump    Six    88  female
10  11       Ronald    Six    89  female
31  32    Binn Rott  Seven    90  female
11  12        Recky    Six    94  female
32  33    Kenn Rein    Six    96  female

In [ ]:
## How many male and female students are there?
df["gender"].value_counts()

gender
male      17
female    16
Name: count, dtype: int64

In [ ]:
## Class and number of student
df["class"].value_counts()

class
Seven    10
Four      8
Six       7
Three     3
Nine      2
Five      2
Fifth     1
Eight     1
Name: count, dtype: int64